cub/agent/agent_batch_memcpy.cuh

```diff
@@ -0,0 +1,967 @@
+
+  // Both `out_start_aligned` and `out_end_aligned` are indices into `out_ptr`.
+  // `out_start_aligned` is the first OutT-aligned memory location after `out_ptr + 4`.
+  // `out_end_aligned` is the last OutT-aligned memory location before `out_end - 4`, where out_end corresponds to one
+  // past the last byte to be copied. Bytes between `[out_start_aligned, out_end_aligned)` will be copied using
+  // uint4. `out_ptr + 4` and `out_end - 4` are used instead of `out_ptr` and `out_end` to avoid `LoadVector`
+  // reading beyond data boundaries.
+  int32_t out_start_aligned = (in_datatype_size + alignment_offset + out_datatype_size - 1) / out_datatype_size *
+                                out_datatype_size -
+                              alignment_offset;
+  int32_t out_end_aligned = (num_bytes - in_datatype_size + alignment_offset) / out_datatype_size * out_datatype_size -
+                            alignment_offset;
+
+  // Perform vectorized copies by loading 4B at the time to populate a `VectorT` and write that out to global memory
+  for (int64_t ichar = out_start_aligned + thread_rank * out_datatype_size; ichar < out_end_aligned;
+       ichar += group_size * out_datatype_size)
+  {
+    VectorT data_in;
+    LoadVector(in_ptr + ichar, data_in);
+    *(out_chars_aligned + (ichar + alignment_offset) / out_datatype_size) = data_in;
+  }
+
+  // Tail logic: copy bytes of the current string outside `[out_start_aligned, out_end_aligned)`.
+  if (out_end_aligned <= out_start_aligned)
+  {
+    // In this case, `[out_start_aligned, out_end_aligned)` is an empty set, and we copy the entire string.
+    for (int32_t ichar = thread_rank; ichar < num_bytes; ichar += group_size)
+    {
+      out_ptr[ichar] = in_ptr[ichar];
+    }
+  }
+  else
+  {
+    // Copy bytes in range `[out_start, out_start_aligned)`.
+    if (thread_rank < out_start_aligned)
+    {
+      out_ptr[thread_rank] = in_ptr[thread_rank];
+    }
+    // Copy bytes in range `[out_end_aligned, num_bytes)`.
+    int32_t ichar = out_end_aligned + thread_rank;
+    if (ichar < num_bytes)
+    {
+      out_ptr[ichar] = in_ptr[ichar];
+    }
+  }
+}
+
+/**
+ * @brief A helper class that allows threads to maintain multiple counters, where the counter that shall be incremented
+ * can be addressed dynamically without incurring register spillage.
+ *
+ * @tparam NUM_ITEMS The number of counters to allocate
+ * @tparam MAX_ITEM_VALUE The maximum count that must be supported.
+ * @tparam PREFER_POW2_BITS Whether the number of bits to dedicate to each counter should be a power-of-two. If enabled,
+ * this allows replacing integer multiplication with a bit-shift in exchange for higher register pressure.
+ * @tparam BackingUnitT The data type that is used to provide the bits of all the counters that shall be allocated.
+ */
+template <uint32_t NUM_ITEMS, uint32_t MAX_ITEM_VALUE, bool PREFER_POW2_BITS, typename BackingUnitT = uint32_t>
```

Comment:
I am slightly worried what  happens if we have a lot of items and a small `BackingUnitT ` type

Reply:
That is, concerned about the performance?

Reply:
Concerned about overflowing a smaller type

Reply:
The data structure should be able to correctly cope with lots of items. That's not its designed use case, though, and, hence, will become inefficient